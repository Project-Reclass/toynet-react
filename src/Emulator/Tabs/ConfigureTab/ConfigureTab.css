.configure {
  padding: 20px;
  padding-top: 0px;
  border-radius: 0 10px 10px 10px;
  border-top: 0;
  position: relative;
  background-color: #454950;
}

.configure::-webkit-scrollbar {
  width: 5px;
  color: #454950;
}

.configure::-webkit-scrollbar-track {
  border-radius: 10px;
}

.configure::-webkit-scrollbar-thumb {
  border-radius: 10px;
}

.network-devices::-webkit-scrollbar {
  width: 10px;
  color: #454950;
}

.network-devices::-webkit-scrollbar-track {
  border-radius: 10px;
}

.network-devices::-webkit-scrollbar-thumb {
  border-radius: 10px;
}

.network-devices {
  display: flex;
  overflow: auto;
  overflow-x: hidden;
}

.max-nodes {
  color: red;
  margin-right: auto;
}
