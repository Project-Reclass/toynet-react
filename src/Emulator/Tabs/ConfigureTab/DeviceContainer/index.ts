import DeviceContainer from './DeviceContainer';
export default DeviceContainer;