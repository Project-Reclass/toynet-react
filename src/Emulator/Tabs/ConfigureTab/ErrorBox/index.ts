import ErrorBox from './ErrorBox';

export { default as ErrorBoxProvider } from './ErrorBoxProvider';
export * from './ErrorBoxProvider';

export default ErrorBox;