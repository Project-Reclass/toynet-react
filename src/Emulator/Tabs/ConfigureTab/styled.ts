import styled from '@emotion/styled';

export const NetworkDevices = styled.div`
  display: flex;
  overflow: auto;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;

  ::-webkit-scrollbar {
    width: 10px;
    color: #454950;
  }

  ::-webkit-scrollbar-track {
    border-radius: 10px;
  }

  ::-webkit-scrollbar-thumb {
    border-radius: 10px;
  }
`;

export const Configure = styled.div`
  padding: 20px;
  padding-top: 0px;
  border-radius: 0 10px 10px 10px;
  border-top: 0;
  position: relative;
  background-color: #454950;
  
  ::-webkit-scrollbar {
    width: 5px;
    color: #454950;
  }

  ::-webkit-scrollbar-track {
    border-radius: 10px;
  }

  ::-webkit-scrollbar-thumb {
    border-radius: 10px;
  }
`;