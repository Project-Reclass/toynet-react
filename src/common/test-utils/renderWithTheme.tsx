import React, { FC } from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '@chakra-ui/core/dist';
import renderer from 'react-test-renderer';

const Wrapper: FC = ({ children }) => (
  <ThemeProvider>{children}</ThemeProvider>
);

export const renderWithTheme = (ui: any) => {
  return render(ui, { wrapper: Wrapper });
};

export const renderTreeWithTheme = (ui: any) => {
  return renderer.create(<Wrapper>{ui}</Wrapper>);
};
