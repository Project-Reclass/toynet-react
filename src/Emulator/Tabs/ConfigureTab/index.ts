import ConfigureTab from './ConfigureTab';
export * from './ConfigureTab';
export default ConfigureTab;