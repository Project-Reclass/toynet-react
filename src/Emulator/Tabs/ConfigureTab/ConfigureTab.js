import React from "react";
import "./ConfigureTab.css";
import DeviceContainer from "./DeviceContainer/DeviceContainer";
import { Button } from "react-bootstrap";

// Sample responses from currently non-existent API
const routers = [
  {
    name: "r1",
    connections: ["s1", "s2"],
  },
];
const switches = [
  {
    name: "s1",
    connections: ["r1", "h1", "h2", "h3"],
  },
  {
    name: "s2",
    connections: [],
  },
];
const hosts = [
  {
    name: "h1",
    connections: ["s1"],
  },
  {
    name: "h2",
    connections: ["s1"],
  },
  {
    name: "h3",
    connections: ["s1"],
  },
  {
    name: "h4",
    connections: [],
  },
];

const Configure = () => {
  return (
    <div className="configure">
      <div className="network-devices">
        <DeviceContainer deviceName="Router" devices={routers} />
        <DeviceContainer deviceName="Switch" devices={switches} />
        <DeviceContainer deviceName="Host" devices={hosts} />
      </div>
      <div className="btn-run-container">
        <Button className="btn-run">Run</Button>
      </div>
    </div>
  );
};

export default Configure;
