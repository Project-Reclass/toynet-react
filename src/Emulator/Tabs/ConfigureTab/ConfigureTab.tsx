import React, { useRef, FC, useCallback, useMemo, useState } from 'react';

import { useTopology, TopologyActions } from 'src/common/hooks/useTopology';

import DeviceContainer from './DeviceContainer';
import './ConfigureTab.css';

const MAX_DEVICES = 10;
const FIVE_SECONDS = 5000;

/**
 * Determines the number of the newly added device
 */
export const getNextNumber = (s: string) => Number(s.slice(1)) + 1;

/**
 * Determines the name of the newly added device
 */
export const getNextDeviceName = (device: Array<{name: string}>, deviceLetter: string) => {
  if (device.length < 1) {
    return `${deviceLetter}1`;
  } else {
    const lastDeviceName = device[device.length - 1].name;
    return `${deviceLetter}${getNextNumber(lastDeviceName)}`;
  }
};

const scrollDeviceContainer = (ref: React.MutableRefObject<HTMLDivElement | null>) => {
  // A slight delay seems to be necessary for the scroll to work properly
  // ref.current.scrollIntoView() on its own does not scroll down enough
  if (ref.current) {
    setTimeout(() => {
      if (ref.current)
        ref.current.scrollIntoView();
    }, 0);
  }
};

const ConfigureTab: FC<{status: string}> = ({ status }) => {
  const { switches, hosts, routers, dispatch } = useTopology(1);
  const [showError, setShowError] = useState(false);

  const routerScrollRef = useRef(null);
  const switchScrollRef = useRef(null);
  const hostScrollRef = useRef(null);

  const toggleErrorMessage = useCallback(() => {
    setShowError(true);
    setTimeout(() => {
      setShowError(false);
    }, FIVE_SECONDS);
  }, []);

  const addDevice = useCallback((type: 'router' | 'host' | 'switch') => {
    const device = type === 'router' ? routers :
      type === 'host' ? hosts : switches;
    const actionType = type === 'router' ? TopologyActions.ADD_ROUTER :
      type === 'host' ? TopologyActions.ADD_HOST : TopologyActions.ADD_SWITCH;
    const deviceRef = type === 'router' ? routerScrollRef :
      type === 'switch' ? switchScrollRef : hostScrollRef;

    return (deviceLetter: string) => {
      if (device.length >= MAX_DEVICES) {
        toggleErrorMessage();
        return;
      }

      scrollDeviceContainer(deviceRef);
      dispatch({
        type: actionType,
        payload: {
          type,
          name: getNextDeviceName(device, deviceLetter),
          connections: [],
        },
      });
    };
  }, [dispatch, hosts, routers, switches, toggleErrorMessage]);

  const addRouter = useMemo(() => addDevice('router'), [addDevice]);
  const addSwitch= useMemo(() => addDevice('switch'), [addDevice]);
  const addHost = useMemo(() => addDevice('host'), [addDevice]);

  return (
    <div className={`configure ${status}`}>
      <div className="network-devices">
        <DeviceContainer
          deviceName="Router"
          devices={routers}
          addDevice={addRouter}
          ref={routerScrollRef}
        />
        <DeviceContainer
          deviceName="Switch"
          devices={switches}
          addDevice={addSwitch}
          ref={switchScrollRef}
        />
        <DeviceContainer
          deviceName="Host"
          devices={hosts}
          addDevice={addHost}
          ref={hostScrollRef}
        />
      </div>
      <div className="btn-run-container">
        <div
          className="max-nodes"
          style={{ display: showError ? 'initial' : 'none' }}
        >
          Each device type can have maximum 10 nodes
        </div>
        <button type="button" className="btn-run">
          Run
        </button>
      </div>
    </div>
  );
};

export default ConfigureTab;
