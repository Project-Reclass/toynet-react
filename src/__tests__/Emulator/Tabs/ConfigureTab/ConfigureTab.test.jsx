import React from 'react';

jest.mock('src/common/api/topology/requests')
import ConfigureTab, { getNextDeviceName, getNextNumber } from 'src/Emulator/Tabs/ConfigureTab';
import { render, act, fireEvent } from '@testing-library/react';
import { renderWithTheme } from 'src/common/test-utils/renderWithTheme';

describe('ConfigureTab helper functions', () => {
  it('should increment numbers correctly', () => {
    const randomInt = Math.floor(Math.random() * 10) + 1;

    expect(getNextNumber(`r${randomInt}`)).toEqual(randomInt + 1);
    expect(getNextNumber(`s${randomInt}`)).toEqual(randomInt + 1);
    expect(getNextNumber(`h${randomInt}`)).toEqual(randomInt + 1);
  });

  it('should increment device names correctly', () => {
    const testRouter = [{ name: "r1", connections: [] }];
    const testSwitch = [{ name: "s1", connections: [] }];
    const testHost = [{ name: "h1", connections: [] }];

    expect(getNextDeviceName(testRouter, "r")).toEqual("r2");
    expect(getNextDeviceName(testSwitch, "s")).toEqual("s2");
    expect(getNextDeviceName(testHost, "h")).toEqual("h2");
  });

  it('should handle empty initial configs correctly', () => {
    const emptyConfig = [];

    expect(getNextDeviceName(emptyConfig, "r")).toEqual("r1");
    expect(getNextDeviceName(emptyConfig, "s")).toEqual("s1");
    expect(getNextDeviceName(emptyConfig, "h")).toEqual("h1");
  });
});

describe('ConfigureTab', ()=> {
  it('should match previous snapshots', () => {
    const { container } = renderWithTheme(<ConfigureTab status={'online'} />);
    expect(container).toMatchSnapshot();
  });
});