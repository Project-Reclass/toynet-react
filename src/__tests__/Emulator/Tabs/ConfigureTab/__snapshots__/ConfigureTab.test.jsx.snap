// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigureTab should match previous snapshots 1`] = `
<div
  className="configure online"
>
  <div
    className="network-devices"
  >
    <div
      className="device-container"
    >
      <div
        className="device-container-name"
      >
        <svg
          className="plus-circle"
          data-testid="plus-icon"
          onClick={[Function]}
        >
          add.svg
        </svg>
        Router
      </div>
      <div
        className="device-container-content"
      >
        <div />
      </div>
    </div>
    <div
      className="device-container"
    >
      <div
        className="device-container-name"
      >
        <svg
          className="plus-circle"
          data-testid="plus-icon"
          onClick={[Function]}
        >
          add.svg
        </svg>
        Switch
      </div>
      <div
        className="device-container-content"
      >
        <div />
      </div>
    </div>
    <div
      className="device-container"
    >
      <div
        className="device-container-name"
      >
        <svg
          className="plus-circle"
          data-testid="plus-icon"
          onClick={[Function]}
        >
          add.svg
        </svg>
        Host
      </div>
      <div
        className="device-container-content"
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;
